Example/Frameworks/Database/Logger.swift

```diff
@@ -27,35 +27,133 @@ import UIKit
+enum Message: Willow.Message {
+    case backupComplete
+    case connectionOpened
+    case sqlQuery(sql: String)
+    case failure(error: Error)
+    case sqlFailure(sql: String, error: Error)
+
+    var name: String {
+        switch self {
+        case .connectionOpened: return "Connection opened"
+        case .backupComplete:   return "Backup complete"
+        case .sqlQuery:         return "SQL Query"
+        case .failure:          return "Error"
+        case .sqlFailure:       return "SQL Error"
         }
     }
 
-    let modifiers: [LogLevel: [LogMessageModifier]] = [.all: [PrefixModifier(), TimestampModifier()]]
-    let queue = DispatchQueue(label: "com.nike.database.logger.queue", qos: .utility)
-    let configuration = LoggerConfiguration(modifiers: modifiers, executionMethod: .asynchronous(queue: queue))
+    var attributes: [String: Any] {
+        var keyPathAttributes: [KeyPath: Any] = [:]
+        let success: Bool
+
+        // Fill in message specific attributes
+        switch self {
+        case .connectionOpened:
+            success = true
+
+        case .backupComplete:
+            success = true
+
+        case let .sqlQuery(sql):
+            keyPathAttributes[.sql] = sql
+            success = true
+
+        case let .failure(error):
+            keyPathAttributes[.errorDescription] = message(forError: error)
+            keyPathAttributes[.errorCode] = code(forError: error)
+            success = false
+
+        case let .sqlFailure(sql, error):
+            keyPathAttributes[.sql] = sql
+            keyPathAttributes[.errorDescription] = message(forError: error)
+            keyPathAttributes[.errorCode] = code(forError: error)
+            success = false
+        }
 
-    return Logger(configuration: configuration)
-}()
+        // Assign attributes that should be present for all messages
+        keyPathAttributes[.frameworkName] = Framework.name
+        keyPathAttributes[.frameworkVersion] = Framework.version
+        keyPathAttributes[.result] = success ? "success" : "failure"
+
+        // Map to the expected types
+        var attributes: [String: Any] = [:]
+        keyPathAttributes.forEach { attributes[$0.key.rawValue] = $0.value }
+
+        return attributes
+    }
+
+    /// Information about this framework.
+    private enum Framework {
+        static let name = "Database"
+
+        static let version: String = {
+            class Version {}
+            return Bundle(for: Version.self).infoDictionary?["CFBundleShortVersionString"] as? String ?? "0.0.0"
+        }()
+    }
+
+    /// Attribute keys this framework uses.
+    private enum KeyPath: String {
+        case sql                = "sql"
+        case errorDescription   = "error_description"
+        case errorCode          = "error_code"
+        case result             = "result"
+        case frameworkName      = "framework_name"
+        case frameworkVersion   = "framework_version"
+    }
```

Comment:
Generally we try to have the nested helper types defined at the top of the type...do we think that it's really worth deviating from that in an enum case?